File: backend/schemas/purchases.py
"""Purchase invoice schemas (Pydantic V2). Proxy over the ERPNext Purchase Invoice DocType."""

from pydantic import BaseModel, Field


class BillLineInput(BaseModel):
    item_code: str = Field(min_length=1, max_length=140)
    qty: float = Field(gt=0)
    rate: float = Field(ge=0)


class PurchaseInvoiceCreate(BaseModel):
    supplier: str = Field(min_length=1)
    items: list[BillLineInput] = Field(min_length=1)
    bill_no: str | None = None
    posting_date: str | None = None
    remarks: str | None = None


class PurchaseInvoiceUpdate(PurchaseInvoiceCreate):
    """Replaces a *posted* supplier bill wholesale.

    A submitted Purchase Invoice is immutable in ERPNext, so this is applied as
    cancel-then-amend: every field is rewritten from this payload exactly as a
    create would write it, and an omitted field is cleared rather than kept.
    """


class BillLine(BaseModel):
    item_code: str
    item_name: str | None = None  # as billed; see InvoiceLine in schemas/sales.py
    qty: float
    rate: float
    amount: float


class PurchaseInvoiceRead(BaseModel):
    id: str
    supplier: str
    posting_date: str | None = None
    due_date: str | None = None
    bill_no: str | None = None
    grand_total: float
    outstanding_amount: float
    status: str
    remarks: str | None = None
    items: list[BillLine] = []
    # See schemas/sales.py for what these three carry — the amendment lineage
    # is identical on both invoice types.
    amended_from: str | None = None
    is_cancelled: bool = False
    is_opening: bool = False

File: frontend/lib/purchases.ts
import { api, encodeId } from "@/lib/http";

export type BillLine = {
  item_code: string;
  qty: number;
  rate: number;
  amount: number;
};

export type PurchaseInvoice = {
  id: string;
  supplier: string;
  posting_date: string | null;
  due_date: string | null;
  bill_no: string | null;
  grand_total: number;
  outstanding_amount: number;
  status: string;
  remarks: string | null;
  items: BillLine[];
  amended_from: string | null;
  is_cancelled: boolean;
  is_opening: boolean;
};

export type PurchaseInput = {
  supplier: string;
  items: { item_code: string; qty: number; rate: number }[];
  bill_no?: string | null;
  posting_date?: string | null;
  remarks?: string | null;
};

export const listPurchases = (
  token: string,
  params: { search?: string; status?: string } = {}
) => api.get<PurchaseInvoice[]>(token, "/purchases", { ...params, limit: 200 });

export const getPurchase = (token: string, id: string) =>
  api.get<PurchaseInvoice>(token, `/purchases/${encodeId(id)}`);

export const createPurchase = (token: string, input: PurchaseInput) =>
  api.post<PurchaseInvoice>(token, "/purchases", input);

/**
 * Correct a posted supplier bill — cancel-then-amend, so the bill number
 * changes. See `updateSale` in lib/sales.ts; the semantics are identical.
 */
export const updatePurchase = (token: string, id: string, input: PurchaseInput) =>
  api.put<PurchaseInvoice>(token, `/purchases/${encodeId(id)}`, input);

/** Why this bill cannot be corrected, or null if it can. Mirrors
 * `PurchaseService._ensure_amendable`; the API refuses independently. */
export function amendBlockedReason(
  bill: PurchaseInvoice
): "opening" | "settled" | null {
  if (bill.is_opening) return "opening";
  if (bill.is_cancelled) return null;
  if (bill.grand_total - bill.outstanding_amount > 0.005) return "settled";
  return null;
}
